### PR TITLE
SYCL Copy Code, main branch (2021.04.12.)

### PR DESCRIPTION
**Pull request by krasznaa**
As I said in #67, I wrote a `vecmem::sycl::copy` class as well. But to make it convenient to use in our tests, I added some more code as well.

Basically, I added `vecmem::get_data(...)` functions for some more types. To be able to express in a convenient way that I want to pass "view types" to the SYCL device code, and not host-only types. This could of course be done without the `vecmem::get_data(...)` functions as well, but I thought this would be a bit more elegant than having to create temporary reference variables in the code explicitly.

The `vecmem::sycl::copy` class is pretty much the same as `vecmem::cuda::copy` and `vecmem::hip::copy`. Though unlike those, it actually does have a state. As it needs to hold on to a `sycl::queue` object to be able to perform the memory copies.

And the thing that I learned during these developments, which may be obvious to everybody else, but it wasn't to me... :stuck_out_tongue:  At first I was trying to implement the test for the SYCL copy functionality like:

```c++
   queue.submit( [ &constants, &inputvec,
                   &outputvecdevice ]( cl::sycl::handler& h ) {

      // Create buffer objects for the input data.
      auto const_data = copy.to( vecmem::get_data( constants ),
                                 device_resource );
      auto input_data = copy.to( vecmem::get_data( inputvec ),
                                 device_resource );

      // Run the kernel.
      h.parallel_for< class LinearTransform2 >(
         cl::sycl::range< 1 >( input_data.m_size ),
         [ constants = vecmem::get_data( const_data ),
           input = vecmem::get_data( input_data ),
           output = vecmem::get_data( outputvecdevice ) ]( cl::sycl::id< 1 > id ) {
```

I.e. I tried to create the buffer objects inside the lambda that I would give to `sycl::queue::submit`. Not realising that in such a setup the buffers would be deleted already before the kernel code would start to run. Which lead to a lot of errors in the code of course. Which is why I ended up adding quite some debug output in this PR. :stuck_out_tongue:

Long story short, one must create "buffer objects" outside of `sycl::queue::submit`. :wink:

### CODE REVIEW

File: sycl/include/vecmem/utils/sycl/copy.hpp [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+   /// queue object itself.
+   ///
+   class copy : public vecmem::copy {
+
+   public:
+      /// Constructor on top of a user-provided queue
+      copy( const queue_wrapper& queue = { "" } );
+
+   protected:
+      /// Perform a memory copy using CUDA
```
Comment (paulgessinger) [inline sycl/include/vecmem/utils/sycl/copy.hpp:30]:
```suggestion
      /// Perform a memory copy using SYCL
```
?

File: core/include/vecmem/containers/impl/jagged_vector_buffer.ipp [hunk 2]
```diff
@@ -119,4 +119,20 @@ namespace vecmem::data {
...
 
-} // namespace vecmem::data
+} // namespace data
+
+   template< typename TYPE >
+   data::jagged_vector_view< TYPE >&
+   get_data( data::jagged_vector_buffer< TYPE >& data ) {
+
+      return data;
+   }
```
Comment (paulgessinger) [inline core/include/vecmem/containers/impl/jagged_vector_buffer.ipp:129]: Ok, maybe I'm missing the point, but these functions exist so that you can explicitly write out the buffer -> view conversion (and use a little template deduction)?

**Review by paulgessinger — COMMENTED** [review]
Looks reasonable to me. 

So the *main* thing here is that the copy object takes the queue, such that the actual copy function interface is the same as for the other backends?

File: tests/sycl/test_sycl_containers.sycl [hunk 3]
```diff
@@ -79,3 +82,80 @@ TEST_F( sycl_containers_test, shared_memory ) {
...
+   // Perform a linear transformation using the vecmem vector helper types.
+   queue.submit( [ &const_data, &input_data,
+                   &outputvecdevice ]( cl::sycl::handler& h ) {
+
+      // Run the kernel.
+      h.parallel_for< class LinearTransform2 >(
+         cl::sycl::range< 1 >( input_data.m_size ),
+         [ constants = vecmem::get_data( const_data ),
+           input = vecmem::get_data( input_data ),
+           output = vecmem::get_data( outputvecdevice ) ]( cl::sycl::id< 1 > id ) {
```
Comment (krasznaa) [inline tests/sycl/test_sycl_containers.sycl:132]: The `vecmem::get_data(...)` functions were meant for this sort of usage. Without them we would have to do the following:

```c++
const vecmem::data::vector_view< int >& const_view = const_data;
const vecmem::data::vector_view< int >& input_view = input_data;
vecmem::data::vector_view< int >& output_view = outputvecdevice;

h.parallel_for< class LinearTransform >( cl::sycl::range< 1 >( input_data.m_size ),
                                                               [ const_view, input_view, output_view ]( cl::sycl::id< 1 > id ) {
```

I just wanted to avoid forcing users to have to write out the exact types in their code, if at all possible...

File: tests/sycl/test_sycl_containers.sycl [hunk 4]
```diff
@@ -79,3 +82,80 @@ TEST_F( sycl_containers_test, shared_memory ) {
...
+            const vecmem::const_device_vector< int > inputvec( input );
+            vecmem::device_vector< int > outputvec( output );
+
+            // Perform the linear transformation.
+            outputvec.at( id ) = inputvec.at( id ) * constantarray.at( 0 ) +
+                                 constantarray.at( 1 );
+            return;
+         } );
+   } );
+   queue.wait(); // This is needed just for the OpenCL CPU driver...
```
Comment (krasznaa) [inline tests/sycl/test_sycl_containers.sycl:151]: It turns out I was incorrect here. :confused: I don't know why this test was/is working correctly for me from inside of Docker with a GPU backend. I suspect that the newer version of oneAPI that I'm using, already supports "asynchronous" copies. But when I use the last oneAPI production release, this `wait()` call must be here for all backends...

File: sycl/include/vecmem/utils/sycl/copy.hpp (hunk 1 above)
Comment (krasznaa) [reply sycl/include/vecmem/utils/sycl/copy.hpp:30]: :stuck_out_tongue:

**Review by krasznaa — COMMENTED** [review]
Thanks for the comments!

File: tests/sycl/test_sycl_containers.sycl (hunk 3 above)
Comment (paulgessinger) [reply tests/sycl/test_sycl_containers.sycl:132]: Yeah, exactly, that's what I was thinking of.

**Comment by krasznaa** [comment]
> So the _main_ thing here is that the copy object takes the queue, such that the actual copy function interface is the same as for the other backends?

This was my idea, yes. As mentioned in #67, I also plan to introduce `vecmem::cuda::async_copy` and similar types as well later on. That one for instance will hold on to a `cudaStream_t` object (pointer) for setting up the copies.

Since these "copy objects" are super lightweight, I thought this would be a good design for abstracting some of the algorithmic code from the way that the "actual copies" are made.
